src/tools/kata-ctl/src/ops/check_ops.rs

```diff
@@ -1,44 +1,198 @@
 // Copyright (c) 2022 Intel Corporation
 //
 // SPDX-License-Identifier: Apache-2.0
-//
+//00
+
+use crate::arch::{x86_64::{check},get_arch_type};
 
-use crate::arch;
-use crate::check;
 use crate::ops::version;
 
 use crate::args::{CheckArgument, CheckSubCommand, IptablesCommand, MetricsCommand};
 
-use anyhow::Result;
+use crate::check::{check_version,run_network_checks};
+
+use anyhow::{anyhow, Result};
+
+use std::fmt::{self,Formatter,Display};
 
 const NAME: &str = "kata-ctl";
 
+#[derive(Debug, Clone, Copy)]
+enum ArchType {
+    x86_64,
+    s390x,
+    aarch64,
+    powerpc64le,
+}
+
+enum UserType {
+    Root,
+    NonRoot,
+}
+
+enum PermissionType {
+    Privileged,
+    NonPrivileged,
+}
+
+type BuiltinCmdFp = fn(args: &str) -> Result<()>;
+
+
+struct CheckItem {
+    name: &'static str,
+    descr: &'static str,
+    fp: BuiltinCmdFp,
+    ut: UserType,
+    arch: ArchType,
+    perm: PermissionType,
+}
+
+static CHECK_LIST: &[CheckItem] = &[
+    CheckItem {
+        name: "check-cpu",
+        descr: "This parameter performs the cpu check",
+        fp: check,
+        ut: UserType::NonRoot,
+        arch: ArchType::x86_64,
+        perm: PermissionType::NonPrivileged,
+    },
+
+    CheckItem {
+        name: "check-network",
+        descr: "This parameter performs network check",
+        fp: run_network_checks,
+        ut: UserType::NonRoot,
+        arch: ArchType::x86_64,
+        perm: PermissionType::NonPrivileged,
+    },
+
+    CheckItem {
+        name: "check-version",
+        descr: "Retrieves latest kata version release",
+        fp: check_version,
+        ut: UserType::NonRoot,
+        arch: ArchType::x86_64,
+        perm: PermissionType::NonPrivileged,
+    },
+
+    CheckItem {
+        name: "check-cpu",
+        descr: "This parameter performs the cpu check",
+        fp: check,
+        ut: UserType::NonRoot,
+        arch: ArchType::s390x,
```

Comment:
This is getting quite complex but rather than having a single list of `CheckItem` checks (`CHECK_LIST`), I think it would be better to have one `CHECK_LIST` per architecture. We can then have the generic code call something like `fn get_checks() -> Result<Vec<CheckItem>>`. That way, the generic code would call `get_checks()` and only get a list of the checks for the current architecture. It could either display the details of that list (for `--list`), or execute the functions in order for the normal mode of operation.

That also implies that `CheckItem`:

- Doesn't need the specify `arch`.
- Does need `fp` to point to a platform-specific check function.

Reply:
@jodh-intel thank you, the lifetimes are really important regarding the scopes.

Reply:
@jodh-intel thank you, the structure is now compacted